relay-server/src/services/projects/cache/service.rs

```diff
@@ -0,0 +1,221 @@
+}
+
+/// A service implementing the [`ProjectCache`] interface.
+pub struct ProjectCacheService {
+    store: ProjectStore,
+    source: ProjectSource,
+    config: Arc<Config>,
+
+    project_update_rx: mpsc::UnboundedReceiver<CompletedFetch>,
+    project_update_tx: mpsc::UnboundedSender<CompletedFetch>,
+
+    project_events_tx: broadcast::Sender<ProjectEvent>,
+}
+
+impl ProjectCacheService {
+    /// Creates a new [`ProjectCacheService`].
+    pub fn new(config: Arc<Config>, source: ProjectSource) -> Self {
+        let (project_update_tx, project_update_rx) = mpsc::unbounded_channel();
+        let project_events_tx = broadcast::channel(PROJECT_EVENTS_CHANNEL_SIZE).0;
+
+        Self {
+            store: ProjectStore::default(),
+            source,
+            config,
+            project_update_rx,
+            project_update_tx,
+            project_events_tx,
+        }
+    }
+
+    /// Consumes and starts a [`ProjectCacheService`].
+    ///
+    /// Returns a [`ProjectCacheHandle`] to access the cache concurrently.
+    pub fn start(self) -> ProjectCacheHandle {
+        let (addr, addr_rx) = relay_system::channel(Self::name());
+
+        let handle = ProjectCacheHandle {
+            shared: self.store.shared(),
+            config: Arc::clone(&self.config),
+            service: addr,
+            project_events: self.project_events_tx.clone(),
+        };
+
+        self.spawn_handler(addr_rx);
+
+        handle
+    }
+
+    /// Schedules a new [`Fetch`] and delivers the result to the [`Self::project_update_tx`] channel.
+    fn schedule_fetch(&self, fetch: Fetch) {
+        let source = self.source.clone();
+        let project_updates = self.project_update_tx.clone();
+
+        tokio::spawn(async move {
+            tokio::time::sleep_until(fetch.when()).await;
+
+            // TODO: cached state for delta fetches, maybe this should just be a revision?
+            let state = match source
+                .fetch(fetch.project_key(), false, fetch.revision())
+                .await
+            {
+                Ok(result) => result,
+                Err(err) => {
+                    relay_log::error!(
+                        tags.project_key = fetch.project_key().as_str(),
+                        error = &err as &dyn std::error::Error,
+                        "failed to fetch project from source: {fetch:?}"
+                    );
+
+                    // TODO: change this to ProjectState::Pending once we consider it safe to do so.
+                    // see https://github.com/getsentry/relay/pull/4140.
+                    ProjectState::Disabled.into()
+                }
+            };
+
+            let _ = project_updates.send(fetch.complete(state));
+        });
+    }
+}
+
+/// All [`ProjectCacheService`] message handlers.
+impl ProjectCacheService {
+    fn handle_fetch(&mut self, project_key: ProjectKey) {
+        if let Some(fetch) = self.store.try_begin_fetch(project_key, &self.config) {
```

Comment:
nit: Even though `try_begin_fetch` does not return a boolean, I would rename it to `should_fetch` to clarify that it does not start fetching immediately.